src/legacy/server/logging/rotate/log_rotator.js

```diff
@@ -0,0 +1,292 @@
+    await this._rotate();
+  }
+
+  _shouldRotate() {
+    // should rotate evaluation
+    // 1. should rotate if current log size exceeds
+    //    the defined one on everyBytes or has already
+    //    pass the defined time on interval.
+    // 2. should not rotate if is already rotating or if any
+    //    of the conditions on 1. do not apply
+    if (this.isRotating) {
+      return false;
+    }
+
+    return this.logFileSize >= this.everyBytes;
+  }
+
+  async _rotate() {
+    if (!this._shouldRotate()) {
+      return;
+    }
+
+    await this._rotateNow();
+  }
+
+  async _rotateNow() {
+    // rotate process
+    // 1. get rotated files metadata (list of log rotated files present on the log folder, numerical sorted)
+    // 2. delete last file
+    // 3. rename all files to the correct index +1
+    // 4. rename + compress current log into 1
+    // 5. send SIGHUP to reload log config
+    if (this.isRotating) {
+      return false;
+    }
+
+    // rotate process is starting
+    this.isRotating = true;
+
+    // get rotated files metadata
+    const rotatedFiles = await this._readRotatedFilesMetadata();
+
+    // delete last file
+    await this._deleteLastRotatedFile(rotatedFiles);
+
+    // rename all files to correct index + 1
+    // and normalize
+    await this._renameRotatedFilesByOne(rotatedFiles);
+
+    // rename + compress current log into 0
+    await this._rotateCurrentLogFile();
+
+    // send SIGHUP to reload log configuration
+    this._sendReloadLogConfigSignal();
+
+    // Reset log file size
+    this.logFileSize = 0;
+
+    // rotate process is finished
+    this.isRotating = false;
+  }
+
+  async _readRotatedFilesMetadata() {
+    const logFileBaseName = basename(this.logFilePath);
+    const logFilesFolder = dirname(this.logFilePath);
+
+    return (await readdirAsync(logFilesFolder))
+      .filter(file => new RegExp(`${logFileBaseName}\\.\\d`).test(file))
+      .sort((a, b) => Number(a.match(/(\d+)/g)[0]) - Number(b.match(/(\d+)/g)[0]))
+      .map(filename => `${logFilesFolder}${sep}${filename}`)
+      .filter(filepath => fs.existsSync(filepath));
+  }
+
+  async _deleteLastRotatedFile(rotatedFiles) {
+    if (rotatedFiles.length < this.keepFiles) {
+      return;
+    }
+
+    const lastFilePath = rotatedFiles.pop();
+    await unlinkAsync(lastFilePath);
+  }
+
+  async _renameRotatedFilesByOne(rotatedFiles) {
+    const logFileBaseName = basename(this.logFilePath);
+    const logFilesFolder = dirname(this.logFilePath);
+
+    for (let i = rotatedFiles.length - 1; i >= 0; i--) {
+      const oldFilePath = rotatedFiles[i];
+      const newFilePath = `${logFilesFolder}${sep}${logFileBaseName}.${i + 1}`;
+      await renameAsync(oldFilePath, newFilePath);
+    }
+  }
+
+  async _rotateCurrentLogFile() {
+    const newFilePath = `${this.logFilePath}.0`;
+    await renameAsync(this.logFilePath, newFilePath);
+  }
+
+  _sendReloadLogConfigSignal() {
+    if (isMaster) {
+      process.emit('SIGHUP');
```

Comment:
I don't think the second part about the tests would be much of a problem, we can just mock it for every test. 

The first part I haven't understood. We are already using the `SIGHUP` to reload the configuration and in order to get this to work for our current requirements I need to get the config reloaded so the legacy logger could re-instantiate and configure a new stream output. That would not be needed in case I was able to apply a truncate and create new strategy but I  for the time being I need to use a rename strategy instead. The need for that strategy appears due to the cloud team requirements. They are using filebeat for consuming the logs and filebeat tracks the inode of the file to avoid duplicates. For this to work it is important that the file rotation renames the current file and creates a new file as replacement, rather than copy and truncate which would result in filebeat thinking the rotated file is a new file.
By using the rename and create pattern, filebeat will track that the file was renamed and can still keep processing it from the same position. This way there is no duplicates and no messages lost.